lib/semantics/resolve-names.cpp

```diff
@@ -5954,20 +5987,135 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
+    const parser::ExecutionPartConstruct &x) {
+  if (auto *y{std::get_if<parser::ExecutableConstruct>(&x.u)}) {
+    if (auto *z{std::get_if<Indirection<parser::DoConstruct>>(&y->u)}) {
+      return &z->value();
+    }
+  }
+  return nullptr;
+}
+
+std::size_t OmpAttributeVisitor::GetAssociatedLoopLevelFromClauses(
+    const parser::OmpClauseList &x) {
+  std::size_t orderedLevel{0};
+  std::size_t collapseLevel{0};
+  for (const auto &clause : x.v) {
+    if (const auto *orderedClause{
+            std::get_if<parser::OmpClause::Ordered>(&clause.u)}) {
+      if (const auto v{
+              GetClauseIntValue<parser::OmpClause::Ordered>(*orderedClause)}) {
+        orderedLevel = *v;
+      }
+    }
+    if (const auto *collapseClause{
+            std::get_if<parser::OmpClause::Collapse>(&clause.u)}) {
+      if (const auto v{GetClauseIntValue<parser::OmpClause::Collapse>(
+              *collapseClause)}) {
+        collapseLevel = *v;
+      }
+    }
+  }
+
+  if (orderedLevel &&
+      (!collapseLevel || (collapseLevel && orderedLevel >= collapseLevel))) {
+    return orderedLevel;
+  } else if (!orderedLevel && collapseLevel) {
+    return collapseLevel;
+  }
+  return 1; // default is outermost loop
+}
+
+// 2.15.1.1 Data-sharing Attribute Rules - Predetermined
+//   - The loop iteration variable(s) in the associated do-loop(s) of a do,
+//     parallel do, taskloop, or distribute construct is (are) private.
+//   - The loop iteration variable in the associated do-loop of a simd construct
+//     with just one associated do-loop is linear with a linear-step that is the
+//     increment of the associated do-loop.
+//   - The loop iteration variables in the associated do-loops of a simd
+//     construct with multiple associated do-loops are lastprivate.
+//
+// This assumes that the do-loops association for collapse/ordered clause has
+// been performed (the number of nested do-loops >= n).
+void OmpAttributeVisitor::PrivatizeAssociatedLoopIndex(
+    const parser::OpenMPLoopConstruct &x) {
+  using Bounds = parser::LoopControl::Bounds;
+  std::size_t level{GetContext().associatedLoopLevel};
+  Symbol::Flag ivDSA{Symbol::Flag::OmpPrivate};
+  if (simdSet.test(GetContext().directive)) {
+    if (level == 1) {
+      ivDSA = Symbol::Flag::OmpLinear;
+    } else {
+      ivDSA = Symbol::Flag::OmpLastPrivate;
+    }
+  }
+
+  auto &outer{std::get<std::optional<parser::DoConstruct>>(x.t)};
+  for (const parser::DoConstruct *loop{&*outer}; loop && level > 0; --level) {
+    // go through all the nested do-loops and resolve index variables
+    auto &loopControl{loop->GetLoopControl().value()};
+    const Bounds &bounds{std::get<Bounds>(loopControl.u)};
+    const parser::Name &iv{bounds.name.thing};
+    auto *symbol{ResolveOmp(iv, ivDSA)};
+    iv.symbol = symbol; // adjust the symbol within region
+    AddToContextObjectWithDSA(*symbol, ivDSA);
+
+    const auto &block{std::get<parser::Block>(loop->t)};
+    const auto it{block.begin()};
```

Comment:
Is it possible to have OMP directives on the inner loops that would break the assumption that the next inner do loop is the first element of the block list of statements ?

Reply:
Yes. The Spec requires that all the associated do-loops should be perfectly nested (no code or OpenMP directive in between). Ideally, the OpenMP structural checker will be able to capture this kind of error prior to Name Resolution.